py/kubeflow/kubeflow/ci/update_jupyter_web_app.py

```diff
@@ -0,0 +1,269 @@
+"""Script to build and update the Jupyter WebApp image.
+
+Requires python3
+
+hub CLI depends on an OAuth token with repo permissions:
+https://hub.github.com/hub.1.html
+  * It will look for environment variable GITHUB_TOKEN
+"""
+
+import logging
+import os
+import re
+import tempfile
+import yaml
+
+import fire
+import git
+import httplib2
+
+from kubeflow.testing import util # pylint: disable=no-name-in-module
+
+from containerregistry.client import docker_creds
+from containerregistry.client import docker_name
+from containerregistry.client.v2_2 import docker_http
+from containerregistry.client.v2_2 import docker_image as v2_2_image
+from containerregistry.transport import transport_pool
+
+class WebAppUpdater(object): # pylint: disable=useless-object-inheritance
+  def __init__(self):
+    self._last_commit = None
+
+  def build_image(self, build_project, registry_project):
+    """Build the image.
+
+    Args:
+      build_project: GCP project used to build the image.
+      registry_project: GCP project used to host the image.
+    """
+    env = dict()
+    env.update(os.environ)
+    env["PROJECT"] = build_project
+    env["REGISTRY_PROJECT"] = registry_project
+
+    with tempfile.NamedTemporaryFile() as hf:
+      name = hf.name
+    env["OUTPUT"] = name
+    web_dir = self._component_dir()
+    util.run(["make", "build-gcb"], env=env, cwd=web_dir)
+
+    with open(name) as hf:
+      data = yaml.load(hf)
+
+    return data["image"]
+
+  def update_prototype(self, image):
+    values = {"image": image}
+
+    regexps = {}
+    for param, value in values.items():
+      r = re.compile(r"([ \t]*" + param + ":+ ?\"?)[^\",]+(\"?,?)")
+      v = r"\g<1>" + value + r"\2"
+      regexps[param] = (r, v, value)
+
+    prototype_file = os.path.join(self._root_dir(),
```

Comment:
Should this be some sort of environment variable for the Kubeflow source root?

Reply:
The assumption is that you are running update_jupyter_web_app.py from the clone of the source repository that contains the source code you want to use. 
So we always build jupyter_web_app at the same version of the code we want to build